### DISCUSSION TITLE
Util class to establish a state management convention for team projects with jotai

### DISCUSSION THREAD

**Discussion by pengooseDev** [pmndrs/jotai#2376] [Ideas]
Hello, Teams!  I'm leaving a discussion to seek opinions before make a PR :)

 In team projects that utilize global state management tools, there often arises a problem where abstraction of state-related code occurs within the View (Component). Such code, written in this manner, results in high coupling between the View and State, making maintenance and scalability difficult.

 From this perspective, Redux is often regarded as "more reliable" in team projects because it enforces the location of action (setter) abstraction within the Redux code itself, thereby decoupling from the component (View) by one layer. However, Redux does not enforce a specific abstraction method for selectors (getters). Therefore, it is generally used only to the extent that folder structure and documentation conventions are regulated in team projects.

 This discussion is not about documentation but rather a contemplation on using abstract classes --that can use optionally-- and Jotai to enforce a convention for state management abstraction. Specifically, it is about utilizing an abstract class to mandate the convention for selectors (getters) and actions (setters) in utility classes. 

---
# AtomManager
```ts
import { Atom, WritableAtom } from 'jotai';
/**
 * AtomManager is an abstract class designed to encapsulate and manage the state logic using Jotai atoms.
 * It provides a structured way to define and access the state (`atom`) and its parts (`selectors`)
 * and actions (`actions`) to manipulate the state. This pattern allows for organizing state-related logic
 *
 * @template T The type of the state managed by this AtomManager.
 */
export declare abstract class AtomManager<T> {
    protected initialState: T;
    protected atom: WritableAtom<T, any, void>;
    constructor(initialState: T);
    /**
     * Abstract property `selectors` should be implemented to return a mapping of functions
     * that utilize Jotai's `get` to read specific parts of the state.
     */
    abstract selectors: {
        [K in keyof T]: Atom<T[K]>;
    };
    /**
     * Abstract property `actions` should be implemented to return a mapping of functions
     * that utilize Jotai's `set` to update the state. These actions can be used to encapsulate
     * state mutations, making state management actions more discoverable and reusable.
     */
    abstract actions: {
        [key: string]: WritableAtom<T | null, any, void>;
    };
}
```

# Example

### Example Data
```ts
// example/types.ts
export interface Music {
  id: string;
  title: string;
  thumbnail: string;
  url: string;
}

export interface PlaylistStatus {
  playlist: Music[];
  index: number;
}
```

## `Playlist` state manager:
```ts
import { atom } from 'jotai';
import { AtomManager } from '@types/atom';
import { PlaylistStatus, Music } from '@/types';

// Extend AtomManager to create state manager
export class Playlist extends AtomManager<PlaylistStatus> {
  // Implement selectors and actions
  public selectors = {
      playlist: atom((get) => {
        const { playlist } = get(this.atom);

        return playlist;
      }),

      index: atom((get) => {
        const { index } = get(this.atom);

        return index;
      }),

      currentMusic: atom((get) => {
        const { playlist, index } = get(this.atom);

        return playlist[index];
      }),
  }

  public actions = {
      add: atom(null, (get, set, music: Music) => {
        const { playlist } = get(this.atom);

        if (playlist.some(({ id }) => id === music.id)) return;

        set(this.atom, (prev: PlaylistStatus) => ({
          ...prev,
          playlist: [...prev.playlist, music],
        }));
      }),

      next: atom(null, (get, set) => {
        const { playlist, index } = get(this.atom);

        const isEmpty = this.isEmpty(playlist);

        if (isEmpty) return;

        const isLastMusic = index === playlist.length - 1;

        set(this.atom, (prev: PlaylistStatus) => ({
          ...prev,
          index: isLastMusic ? 0 : prev.index + 1,
        }));
      }),
  }

  // This approach also allows for higher cohesion by encapsulating utility methods related to the specific domain as private, which are used in the context of the statement.
  private isEmpty(playlist: Music[]) {
    return playlist.length === 0;
  }

  private isFirstMusic(index: number) {
    return index === 0;
  }
}

// Create an instance of the state manager
const initialState: PlaylistStatus = {
  playlist: [],
  index: 0,
};

export const playlistManager = new Playlist(initialState);
```

---

## Best Flow

The `AtomManager` class is designed to be used with custom hooks to encapsulate the state management logic and make it easier to use in your components.

> Flow: Class(AtomManager) --> custom hook --> Component(View)

### Custom Hook

```tsx
// usePlaylist.ts
import { useAtomValue, useSetAtom } from 'jotai';
import { playlistManager } from '@/viewModel';

export const usePlaylist = () => {
  const {
    selectors: { playlist, currentMusic },
    actions: { play, next, prev, add, remove },
  } = playlistManager;

  
  // It seems possible to improve by introducing a higher-order function in the future.(duplicated useAtomValue, useSetAtom)
  return {
    // Getters(Selectors)
    playlist: useAtomValue(playlist),
    currentMusic: useAtomValue(currentMusic),

    // Setters(Actions)
    play: useSetAtom(play),
    next: useSetAtom(next),
    prev: useSetAtom(prev),
    add: useSetAtom(add),
    remove: useSetAtom(remove),
  };
};
```

```tsx
// Playlist.tsx
import { usePlaylist } from '@/hooks';
import { Music } from '@/types';

export const Playlist = () => {
  const { playlist, currentMusic, play, next, prev, add, remove } =
    usePlaylist();

  return (
    <div>
      <h1>Playlist</h1>
      <ul>
        {playlist?.map((music: Music[]) => {
          const { id, title, thumbnail } = music;

          return (
            <li key={id}>
              <img src={thumbnail} alt={title} />
              <p>{title}</p>
              <button onClick={() => add(music)}>Add to Playlist</button>
              <button onClick={() => remove(music)}>Remove</button>
              <button onClick={() => play(music)}>Play</button>
            </li>
          );
        })}
      </ul>
      <button onClick={prev}>Prev</button>
      <button onClick={next}>Next</button>
    </div>
  );
};
```

 By providing this as a utility class within the library, and by providing documentation on its usage, we can collaboratively improve the code. By using the specified abstract class as needed, I believe we can enhance both the flexibility of Jotai and the stability of state conventions for users.
 I would like to hear your thoughts on this! ☺️

**Comment by dai-shi** [comment]
Hi, thanks for opening a discussion.

While I'm not very familiar with team projects, such a util aren't very Jotai way, and it might not be for many users. So, I would consider it an open area for third-party libraries.
I'd suggest you to start with a new library under your profile and get some feedback.
We can host it under http://github.com/jotaijs if you want to maintain it in a long term.

**Reply by pengooseDev** [reply]
As you said, we will introduce it within the team project first, and then we will improve it further. Thank you for your time! 😆❤

**Comment by rothsandro** [comment]
The `Playlist` class implements `selectors` and `actions` as getters. This will generate a new atom for each usage of the hook and on every render, resulting in a endless loop (and avoids caching the computed values).

The `usePlaylist` hook subscribes to all atoms. If a component only needs a single state (e.g. `currentMusic`), it still subscribes to all atoms and re-renders unnecessarily when a unused state changes.

Separating state handling code from the view is definitely a good idea but the abstract class also limits the way you can work with atoms (like working with multiple writable atoms, reducers, storage, etc.). You either extend the atom manager to support these cases, making it more complex to maintain, or you go without the atom manager in such cases, making the implementation / convention less useful.

**Reply by pengooseDev** [reply]
Thank you for your great advice! :) I'll think more about it and develop it based on what you said. Thank you! 😀👍
